runtime/parachains/src/scheduler.rs

```diff
@@ -576,54 +576,27 @@ impl<T: Config> Module<T> {
 		Some(GroupIndex(group_idx as u32))
 	}
 
-	/// Returns an optional predicate that should be used for timing out occupied cores.
+	/// Returns a predicate that should be used for timing out occupied cores.
 	///
-	/// If `None`, no timing-out should be done. The predicate accepts the index of the core, and the
+	/// The predicate accepts the index of the core, and the
 	/// block number since which it has been occupied, and the respective parachain and parathread
 	/// timeouts, i.e. only within `max(config.chain_availability_period, config.thread_availability_period)`
 	/// of the last rotation would this return `Some`, unless there are no rotations.
-	///
-	/// This really should not be a box, but is working around a compiler limitation filed here:
-	/// https://github.com/rust-lang/rust/issues/73226
-	/// which prevents us from testing the code if using `impl Trait`.
-	pub(crate) fn availability_timeout_predicate() -> Option<Box<dyn Fn(CoreIndex, T::BlockNumber) -> bool>> {
+	pub(crate) fn availability_timeout_predicate() -> impl Fn(CoreIndex, T::BlockNumber) -> bool {
 		let now = <frame_system::Pallet<T>>::block_number();
 		let config = <configuration::Module<T>>::config();
 
-		let session_start = <SessionStartBlock<T>>::get();
-		let blocks_since_session_start = now.saturating_sub(session_start);
-		let blocks_since_last_rotation = blocks_since_session_start % config.group_rotation_frequency;
-
-		let absolute_cutoff = sp_std::cmp::max(
-			config.chain_availability_period,
-			config.thread_availability_period,
-		);
-
 		let availability_cores = AvailabilityCores::get();
 
-		if blocks_since_last_rotation >= absolute_cutoff {
-			None
-		} else {
-			Some(Box::new(move |core_index: CoreIndex, pending_since| {
-				match availability_cores.get(core_index.0 as usize) {
-					None => true, // out-of-bounds, doesn't really matter what is returned.
-					Some(None) => true, // core not occupied, still doesn't really matter.
-					Some(Some(CoreOccupied::Parachain)) => {
-						if blocks_since_last_rotation >= config.chain_availability_period {
-							false // no pruning except recently after rotation.
-						} else {
-							now.saturating_sub(pending_since) >= config.chain_availability_period
-						}
-					}
-					Some(Some(CoreOccupied::Parathread(_))) => {
-						if blocks_since_last_rotation >= config.thread_availability_period {
-							false // no pruning except recently after rotation.
-						} else {
-							now.saturating_sub(pending_since) >= config.thread_availability_period
-						}
-					}
-				}
-			}))
+		move |core_index: CoreIndex, pending_since| {
+			match availability_cores.get(core_index.0 as usize) {
+				None => true, // out-of-bounds, doesn't really matter what is returned.
+				Some(None) => true, // core not occupied, still doesn't really matter.
+				Some(Some(CoreOccupied::Parachain)) =>
+					now.saturating_sub(pending_since) >= config.chain_availability_period,
```

Comment:
I think this alters behaviour quite a bit:

Before: Only within `rot..(rot + Δ)` the generated `Fn` could return `true` (where `rot` is the time of the _last_ rotation).

After: The only within `rot..(rot + Δ)` is removed.

As you stated, this clears the pending ones way more often and hence could have the desired effect.